### DISCUSSION TITLE
Enable Link-Time Optimization (LTO) and other optimizations

### DISCUSSION THREAD

**Discussion by zamazan4ik** [pendulum-project/ntpd-rs#1522] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) is not enabled. I suggest enabling it since it will reduce the binary size (always a good thing to have) and highly likely will improve the application's performance.

I suggest enabling LTO only for the Release builds to not sacrifice the developers' experience during the working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `release-lto` profile where additionally to regular `release` optimizations LTO also will be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of `ntpd-rs`.

According to my quick local tests, I've achieved the following binary size reduction with just enabling `lto = true` for the Release build profile (first number - without LTO, the second - with LTO):

* ntp-ctl: 57 Mib to 19 Mib
* ntp-daemon: 66 Mib to 42 Mib
* ntp-metrics-exporter: 57 Mib to 20 Mib
* nts-pool-ke: 51 Mib to 30 Mib

After applying LTO I can suggest evaluating other optimization options like evaluating Profile-Guided Optimization (PGO) (more materials and benchmarks are available at my repo: https://github.com/zamazan4ik/awesome-pgo) and Post-Link Optimization (PLO) with tools like LLVM BOLT.

Thank you.

**Comment by davidv1992** [comment]
Thank you for the suggestion. Given the size improvements, this is looking really interesting, and I have created an issue for enabling LTO. Going beyond that is something we'll need to evaluate more, but we'll have a look  at those as well.
